fix: Remove cover of product line item if media has been deleted

aragon999 opened the pull request:
### 1. Why is this change necessary?
When a media is deleted, which has been assigned to a product, which is in an existing cart, the order process will fail:
```
An exception occurred while executing a query: SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`xxx`.`order_line_item`, CONSTRAINT `fk.order_line_item.cover_id` FOREIGN KEY (`cover_id`) REFERENCES `media` (`id`) ON DELETE SET NULL ON UPDATE CASCADE)
```

### 2. What does this change do, exactly?
Set the cover to `null` if the media is not existing on the product anymore.

### 3. Describe each step to reproduce the issue or behaviour.
Put an item into the cart, with an cover, remove the media of the cover item, and try to complete the order.

### 4. Please link to the relevant issues (if any).
\-

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

aragon999 committed:
fix: Remove cover of product line item if media has been deleted

mitelg commented on changelog/_unreleased/2024-08-05-remove-cover-of-line-item-if-media-has-been-deleted.md:
```diff
@@ -0,0 +1,9 @@
+author_github: @aragon999
+---
+# Core
+* Changed `ProductCartProcessor` to remove the cover of a product if the media has been deleted, which would otherwise result in an error if the customer tries to perform the orde` to remove the cover of a product if the media has been deleted, which would otherwise result in an error if the customer tries to perform the order
```
```suggestion
* Changed `ProductCartProcessor` to remove the cover of a product if the media has been deleted, which would otherwise result in an error if the customer tries to perform the order
```

aragon999 replied:
Oopsie :see_no_evil:

mitelg commented on tests/unit/Core/Checkout/Cart/ProductCartProcessorTest.php:
```diff
@@ -170,4 +171,42 @@ public function testPayloadIsTranslated(): void
+        $data = new CartDataCollection();
+        $data->set('product-A', $product);
+
+        static::assertInstanceOf(MediaEntity::class, $cart->getLineItems()->get('A')->getCover());
```
```suggestion
        static::assertInstanceOf(MediaEntity::class, $cart->getLineItems()?->get('A')->getCover());
```

this should fix PHPStan

aragon999 replied:
Thank you :-)

mitelg commented on tests/unit/Core/Checkout/Cart/ProductCartProcessorTest.php:
```diff
@@ -170,4 +171,42 @@ public function testPayloadIsTranslated(): void
+
+        $processor->collect($data, $cart, $context, new CartBehavior());
+
+        static::assertNull($cart->getLineItems()->get('A')->getCover());
```
here we should first check, that the line item is not null and then get the cover. as it would be a falsy test, if 'A' is not a LineItem

aragon999 replied:
Should be fixed as well :-)

aragon999 commented:
@mitelg I just rebased the PR and fixed the merge conflict. Is it possible that someone has a look at this, as nothing happened here since August? :-)

leduc92 commented:
Hi @aragon999, Thanks for your contribution; you did a lot for us. 💙  Your merge request has been merged.
